packages/contracts/contracts/EpochKeyProofVerifier.sol

```diff
@@ -0,0 +1,79 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import {IUnirep} from './interfaces/IUnirep.sol';
+import {IVerifier} from './interfaces/IVerifier.sol';
+
+contract EpochKeyProofVerifier is IUnirep {
+    IVerifier public immutable epochKeyVerifier;
+
+    uint256 public constant SNARK_SCALAR_FIELD =
+        21888242871839275222246405745257275088548364400416034343698204186575808495617;
+
+    constructor(IVerifier _epochKeyVerifier) {
+        epochKeyVerifier = _epochKeyVerifier;
+    }
+
+    function decodeEpochKeyControl(
+        uint256 control
+    )
+        public
+        pure
+        returns (
+            uint256 revealNonce,
+            uint256 attesterId,
+            uint256 epoch,
+            uint256 nonce
+        )
+    {
+        revealNonce = (control >> 232) & 1;
+        attesterId = (control >> 72) & ((1 << 160) - 1);
+        epoch = (control >> 8) & ((1 << 64) - 1);
+        nonce = control & ((1 << 8) - 1);
+        return (revealNonce, attesterId, epoch, nonce);
+    }
+
+    function decodeEpochKeySignals(
+        uint256[] calldata publicSignals
+    ) public pure returns (EpochKeySignals memory) {
+        EpochKeySignals memory signals;
+        signals.epochKey = publicSignals[0];
+        signals.stateTreeRoot = publicSignals[1];
+        signals.data = publicSignals[3];
+        // now decode the control values
+        (
+            signals.revealNonce,
+            signals.attesterId,
+            signals.epoch,
+            signals.nonce
+        ) = decodeEpochKeyControl(publicSignals[2]);
+        return signals;
+    }
+
+    function verifyEpochKeyProof(
```

Comment:
It might be good to include an additional function that checks the `attesterId` variable. This could be like `verifyAndCheckCaller`. This would perform all the checks here, but also have `require(attesterId == msg.sender)`.

This should be a separate function because the `msg.sender` will not _always_ be the attester.

Reply:
Should this be the function that is called by default? Also, should I add tests for this?

Reply:
I don't think either of them need to be called by default. Either can be used by someone that wants to verify.

To test you should generate a proof with an attester id that is a contract address you control, e.g.

```js
const [owner] = await ethers.getAccounts()
```

You'd make a proof with `owner.address` as the attester id and make sure it doesn't revert (when you called `verifyAndCheckCaller` from `owner`. Then make another proof with a random attester id and make sure that it reverts when called with `owner`.